core/chains/evm/logpoller/parser_test.go

```diff
@@ -0,0 +1,339 @@
+
+			assertArgs(t, args, 1)
+		})
+
+		t.Run("unconfirmed", func(t *testing.T) {
+			parser := &pgDSLParser{}
+			chainID := big.NewInt(1)
+			expressions := []query.Expression{query.Confirmation(primitives.Unconfirmed)}
+			limiter := query.LimitAndSort{}
+
+			result, args, err := parser.buildQuery(chainID, expressions, limiter)
+			expected := "SELECT evm.logs.* " +
+				"FROM evm.logs " +
+				"WHERE evm_chain_id = :evm_chain_id " +
+				"AND block_number <= (SELECT greatest(block_number - :confs_0, 0) FROM evm.log_poller_blocks WHERE evm_chain_id = :evm_chain_id ORDER BY block_number DESC LIMIT 1)"
+
+			require.NoError(t, err)
+			assert.Equal(t, expected, result)
+
+			assertArgs(t, args, 2)
+		})
+	})
+
+	t.Run("query for event by word", func(t *testing.T) {
+		t.Parallel()
+
+		wordFilter := NewEventByWordFilter(common.HexToHash("0x42"), 8, []primitives.ValueComparator{
+			{Value: "", Operator: primitives.Gt},
+		})
+
+		parser := &pgDSLParser{}
+		chainID := big.NewInt(1)
+		expressions := []query.Expression{wordFilter}
+		limiter := query.LimitAndSort{}
+
+		result, args, err := parser.buildQuery(chainID, expressions, limiter)
+		expected := "SELECT evm.logs.* " +
+			"FROM evm.logs " +
+			"WHERE evm_chain_id = :evm_chain_id " +
+			"AND substring(data from 32*:word_index_0+1 for 32) > :word_value_0"
+
+		require.NoError(t, err)
+		assert.Equal(t, expected, result)
+
+		assertArgs(t, args, 3)
+	})
+
+	t.Run("query for event topic", func(t *testing.T) {
+		t.Parallel()
+
+		topicFilter := NewEventByTopicFilter(common.HexToHash("0x42"), 8, []primitives.ValueComparator{
+			{Value: "a", Operator: primitives.Gt},
+			{Value: "b", Operator: primitives.Lt},
+		})
+
+		parser := &pgDSLParser{}
+		chainID := big.NewInt(1)
+		expressions := []query.Expression{topicFilter}
+		limiter := query.LimitAndSort{}
+
+		result, args, err := parser.buildQuery(chainID, expressions, limiter)
+		expected := "SELECT evm.logs.* " +
+			"FROM evm.logs " +
+			"WHERE evm_chain_id = :evm_chain_id " +
+			"AND topics[:topic_index_0] > :topic_value_0 AND topics[:topic_index_0] < :topic_value_1"
+
+		require.NoError(t, err)
+		assert.Equal(t, expected, result)
+
+		assertArgs(t, args, 4)
+	})
+
+	// nested query -> a & (b || c)
+	t.Run("nested query", func(t *testing.T) {
+		t.Parallel()
+
+		parser := &pgDSLParser{}
+		chainID := big.NewInt(1)
+		expressions := []query.Expression{
+			{BoolExpression: query.BoolExpression{
+				Expressions: []query.Expression{
+					query.Timestamp(10, primitives.Gte),
+					{BoolExpression: query.BoolExpression{
+						Expressions: []query.Expression{
+							query.TxHash(common.HexToHash("0x84").Hex()),
+							query.Confirmation(primitives.Unconfirmed),
+						},
+						BoolOperator: query.OR,
+					}},
+				},
+				BoolOperator: query.AND,
+			}},
+		}
+		limiter := query.LimitAndSort{}
+
+		result, args, err := parser.buildQuery(chainID, expressions, limiter)
+		expected := "SELECT evm.logs.* " +
```

Comment:
It's very hard to understand when parentheses open and close, can we have it formatted in a more user-friendly way in tests? Sth like this for instance (similar to Go code indentation)
```sql
SELECT evm.logs.*
FROM evm.logs
WHERE evm_chain_id = :evm_chain_id
  AND (
        block_timestamp >= :block_timestamp_0
        AND (
                    tx_hash = :tx_hash_0
                    OR block_number <= (SELECT greatest(block_number - :confs_0, 0)
                                        FROM evm.log_poller_blocks
                                        WHERE evm_chain_id = :evm_chain_id
                                        ORDER BY block_number DESC
                                        LIMIT 1)
                )
    )

```

Reply:
Yes, it is, but spaces included in a formatted string like this compared to the actual SQL string are different and cause the tests to fail. The solutions for that are to remove spaces from one or add spaces to the other.